saleor/dashboard/product/views.py

```diff
@@ -2,66 +2,83 @@
     else:
-        product = get_product_cls_by_name(cls_name)()
-        title = _('Add new %s') % cls_name
+        product = get_object_or_404(
+            Product.objects.select_subclasses().prefetch_related(
+                'images', 'variants'), pk=pk)
+        title = product.name
+    attributes = product.attributes.all()
+    images = product.images.all()
+    variants = product.variants.select_subclasses()
+    stock_items = Stock.objects.filter(variant__in=variants)
+
     form_cls = get_product_form(product)
-    variant_formset_cls = get_variant_formset(product)
-    form = form_cls(request.POST or None, instance=product)
-    variant_formset = variant_formset_cls(
-        request.POST or None, instance=product)
-    image_formset = ProductImageFormSet(
-        request.POST or None, request.FILES or None, instance=product)
-    forms = {'form': form, 'variant_formset': variant_formset,
-             'image_formset': image_formset}
-    if all([f.is_valid() for f in forms.values()]):
-        with transaction.atomic():
-            product = form.save()
-            variant_formset.save()
-            image_formset.save()
-        if pk:
-            msg = _('Updated product %s') % product
-        else:
+    form = form_cls(instance=product, initial=initial)
+
+    if 'product-form' in request.POST:
+        form = form_cls(request.POST, instance=product)
+        product = save_product_form(request, form, creating)
+        if creating and product:
+            return redirect('dashboard:product-update', pk=product.pk)
+
+    ctx = {'title': title, 'product': product, 'attributes': attributes,
+           'images': images, 'product_form': form, 'stock_items': stock_items,
+           'variants': variants}
+    if pk:
+        images_reorder_url = reverse_lazy('dashboard:product-images-reorder',
+                                          kwargs={'product_pk': pk})
+        ctx['images_reorder_url'] = images_reorder_url
+    return TemplateResponse(request, 'dashboard/product/product_form.html', ctx)
+
+
+def save_product_form(request, form, creating):
+    product = None
+    if form.is_valid():
+        product = form.save()
+        if creating:
             msg = _('Added product %s') % product
+        else:
+            msg = _('Updated product %s') % product
         messages.success(request, msg)
-        return redirect('dashboard:products')
     else:
-        if any([f.errors for f in forms.values()]):
+        if form.errors:
             messages.error(request, _('Your submitted data was not valid - '
-                           'please correct the errors below'))
-    ctx = {'title': title, 'product': product}
-    ctx.update(forms)
-    return TemplateResponse(request, 'dashboard/product/product_form.html', ctx)
+                                      'please correct the errors below'))
+    return product
 
 
 class ProductDeleteView(StaffMemberOnlyMixin, DeleteView):
```

Comment:
Can we make this a regular function to match the views for deleting variants and stock?